ledger-app-fio/src/signTransaction.c

```diff
@@ -0,0 +1,796 @@
+
+		sha_256_append(&ctx->hashContext, (uint8_t *) &ctx->amount, SIZEOF(ctx->amount));
+		sha_256_append(&ctx->hashContext, (uint8_t *) &ctx->maxFee, SIZEOF(ctx->maxFee));
+		sha_256_append(&ctx->hashContext, (uint8_t *) wireData2->actor, SIZEOF(wireData2->actor));
+		sha_256_append(&ctx->hashContext, (uint8_t *) wireData2->tpidLength, SIZEOF(wireData2->tpidLength));
+		sha_256_append(&ctx->hashContext, (uint8_t *) wireData2->tpid, wireData2->tpidLength[0]);
+	}
+
+	security_policy_t policy = policyForSignTxActionData(ctx->actionValidationActor, ctx->actionDataActor);
+	TRACE("Policy: %d", (int) policy);
+	ENSURE_NOT_DENIED(policy);
+	{
+		// select UI steps
+		switch (policy) {
+#	define  CASE(POLICY, UI_STEP) case POLICY: {ctx->ui_step=UI_STEP; break;}
+			CASE(POLICY_SHOW_BEFORE_RESPONSE, HANDLE_ACTION_DATA_STEP_SHOW_PUBKEY);
+#	undef   CASE
+		default:
+			THROW(ERR_NOT_IMPLEMENTED);
+		}
+	}
+
+	signTx_handleActionData_ui_runStep();
+}
+
+// ============================== WITNESS ==============================
+
+enum {
+	HANDLE_WITTNESSES_STEP_DISPLAY_DETAILS = 1000,
+	HANDLE_WITTNESSES_STEP_CONFIRM,
+	HANDLE_WITTNESSES_STEP_RESPOND,
+	HANDLE_WITTNESSES_STEP_INVALID,
+} ;
+
+static void signTx_handleWitnesses_ui_runStep()
+{
+	TRACE("UI step %d", ctx->ui_step);
+	TRACE_STACK_USAGE();
+	ui_callback_fn_t* this_fn = signTx_handleWitnesses_ui_runStep;
+
+	UI_STEP_BEGIN(ctx->ui_step, this_fn);
+
+	UI_STEP(HANDLE_WITTNESSES_STEP_DISPLAY_DETAILS) {
+		ui_displayPathScreen(
+		        "Sign with:",
+		        &ctx->path,
+		        this_fn
+		);
+	}
+
+	UI_STEP(HANDLE_WITTNESSES_STEP_CONFIRM) {
+		ui_displayPrompt(
+		        "Sign",
+		        "transaction?",
+		        this_fn,
+		        respond_with_user_reject
+		);
+	}
+
+	UI_STEP(HANDLE_WITTNESSES_STEP_RESPOND) {
+		io_send_buf(SUCCESS, G_io_apdu_buffer, 65 + 32);
+		ui_displayBusy(); // needs to happen after I/O
+		advanceStage();
+	}
+
+	UI_STEP_END(HANDLE_WITTNESSES_STEP_INVALID);
+}
+
+__noinline_due_to_stack__
+void signTx_handleWitnessesAPDU(uint8_t p2, uint8_t* wireDataBuffer, size_t wireDataSize)
+{
+	TRACE_STACK_USAGE();
+	{
+		// sanity checks
+		CHECK_STAGE(SIGN_STAGE_WITNESSES);
+		VALIDATE(p2 == P2_UNUSED, ERR_INVALID_REQUEST_PARAMETERS);
+		ASSERT(wireDataSize < BUFFER_SIZE_PARANOIA);
+	}
+
+	explicit_bzero(&ctx->path, SIZEOF(ctx->path));
+
+	{
+		// parse
+		TRACE_BUFFER(wireDataBuffer, wireDataSize);
+
+		size_t parsedSize = bip44_parseFromWire(&ctx->path, wireDataBuffer, wireDataSize);
+		VALIDATE(parsedSize == wireDataSize, ERR_INVALID_DATA);
+	}
+
+	security_policy_t policy = POLICY_DENY;
+	{
+		// get policy
+		policy = policyForSignTxWitnesses(&ctx->path);
+		TRACE("Policy: %d", (int) policy);
+		ENSURE_NOT_DENIED(policy);
+	}
+
+	//Extension points
+	uint8_t buf[1];
+	explicit_bzero(buf, SIZEOF(buf));
+	sha_256_append(&ctx->hashContext, buf, SIZEOF(buf));
+
+	//We finish the hash appending a 32-byte empty buffer
```

Comment:
I'd consider finishing the tx hash sooner and returning it --- it can then be checked by the client to verify that tx serialization is consistent. Any change / bug is then discovered immediately and not just by the blockchain reporting invalid signatures, which is quite useful if the tx format changes over time.